pkg/agent/ipassigner/virtual/ip_assigner_linux.go

```diff
@@ -0,0 +1,357 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package virtual
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+
+	"antrea.io/antrea/pkg/agent/ipassigner"
+)
+
+const (
+	minRetryDelay                       = 1 * time.Second
+	maxRetryDelay                       = 300 * time.Second
+	solicitedNodeMulticastAddressPrefix = "ff02::1:ff00:0"
+)
+
+// ipAssigner manages external IPs on the Node. It will create raw sockets and listen for ARP requests packets (IPv4)
+// and Neighbor Discovery packets (IPv6) directly without assigning any IP to any interfaces. This avoids the kernel
+// creating local routes for the external IP, which have a higher priority than the routes installed by antrea-proxy
+// in proxyAll mode.
+type ipAssigner struct {
+	arpResponders   map[int]*ARPResponder
+	ndpResponders   map[int]*NDPResponder
+	respondersLock  *sync.Mutex
+	assignedIPs     *sync.Map
+	ipChan          chan net.IP
+	stopChan        chan struct{}
+	linkDetector    ipassigner.LinkDetector
+	multicastGroups map[int]int
+	queue           workqueue.RateLimitingInterface
+}
+
+// NewIPAssigner returns an *ipAssigner.
+func NewIPAssigner() (*ipAssigner, error) {
+	assigner := &ipAssigner{
+		arpResponders:   make(map[int]*ARPResponder),
+		ndpResponders:   make(map[int]*NDPResponder),
+		multicastGroups: make(map[int]int),
+		assignedIPs:     &sync.Map{},
+		respondersLock:  &sync.Mutex{},
+		ipChan:          make(chan net.IP),
+		stopChan:        make(chan struct{}),
+		linkDetector:    ipassigner.NewLinkDetector(),
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "virtualIPAssigner"),
+	}
+	assigner.linkDetector.AddEventHandler(assigner.onLinkUpdate)
+	return assigner, nil
+}
+
+func (a *ipAssigner) hasIP(ip net.IP) bool {
+	_, exist := a.assignedIPs.Load(ip.String())
+	return exist
+}
+
+func (a *ipAssigner) onLinkUpdate(linkIndex int) {
+	a.queue.Add(linkIndex)
```

Comment:
Sorry, I didn't know supporting ARP responder for multiple interfaces requires so many extra work, thought it's just an argument when creating the socket. Creating two responder per interface dynamically seems a bit complex and not worthy as each Pod will have an interface, which means any Pod lifecycle event will trigger it to do some work.
Could we switch to the original simple approach for now? If multiple interfaces is required in the future, I think having a static configuration to take the list of interfaces is also better than supporting all interfaces.

Reply:
For e2e test, maybe it could create a Local externalTrafficPolicy service and access the externalIP from the Node that doesn't have its endpoint?